torch/csrc/jit/tensorexpr/external_functions.cpp

```diff
@@ -196,6 +204,29 @@ void nnc_aten_mean(
   }
 }
 
+void nnc_aten_addmm(
+    int64_t bufs_num,
+    void** buf_data,
+    int64_t* buf_ranks,
+    int64_t* buf_dims,
+    int8_t* buf_dtypes,
+    int64_t args_num,
+    int64_t* extra_args) {
+  std::vector<at::Tensor> tensors =
+      constructTensors(bufs_num, buf_data, buf_ranks, buf_dims, buf_dtypes);
+
+  at::Tensor& r = tensors[0];
+  const at::Tensor& x = tensors[1];
+  const at::Tensor& y = tensors[2];
+  const at::Tensor& z = tensors[3];
+  try {
+    at::addmm_out(r, x, y, z, extra_args[0], extra_args[1]);
+  } catch (...) {
```

Comment:
The idea you were talking about in todays meeting would be to start with a simpler generic try/catch where these (aten based) external functions get called which would then raise the proper exception and debug handle.
Is that right?

Reply:
That's my original thought. @ZolotukhinM mentioned the exception handling is disabled in the generated assembly code, so I think the options are:
1) Figure out how to add exception handling capability in the generated llvm assembly - if I understand correctly just to pass through the exceptions thrown by these methods requires some change;
2) (suggested by @bertmaher ) make these functions to return error code / debug handle instead of throwing; We need codegen return-code check logic (and jump to the end of the function if error code is set) on the callsites of the assembly code. It's basically an ad-hoc exception handling.

Reply:
So if external calls are not in try catch block do we just terminate?

Reply:
So I did not quite understand the reason behind error handling mechanism in tensorexpr. But in TVM, which I guess is larger in scope, there is some dynamic error handling it seem. I did not understand fully but, presumably you can have exprs that have condition checks and they can throw if they fail.

When we have dynamic shape enabled I assume we will need to handle at least the shapes?

Some code from TVM that I am sure you guys know already:
https://codebrowser.bddppq.com/tvm/tvm/include/tvm/tir/stmt.h.html#tvm::tir::AssertStmtNode
https://codebrowser.bddppq.com/tvm/tvm/src/target/llvm/codegen_cpu.cc.html#911

cc: @ZolotukhinM @bertmaher

Reply:
> So I did not quite understand the reason behind error handling mechanism in tensorexpr

Currently, there is no error handling mechanism in TE :) The reason why calls are guarded with catch-all blocks are not fundamental, that's a detail of the current implementation. Eventually we would like to add a support for error handling and as @ljk53 mentioned in the comment above we have two options for that: either 1) properly support exceptions in the generated code, or 2) go with the return code mechanism and assume that exceptions could be raised at the call-site of generated code (where we invoke NNC kernel). From implementation perspective, (2) seems much simpler and if the difference is not critical otherwise, I'd go that route.

Reply:
> So if external calls are not in try catch block do we just terminate?

That's probably an undefined behavior or something: mark the wrappers (e.g. "nnc_aten_addmm") as `no-throw` in the generated LLVM code. I guess if that contract is violated random things could happen as LLVM would optimize the rest of the generated code under that assumption.